### PR TITLE
Introduce vhost crate to support vhost-kernel and vhost-user backend drivers

### PR DESCRIPTION
**Pull request by jiangliu**
This PR migrates vhost_backend and vhost_gen code from the firecracker project into the vhost crate.
Support of vhost-user protocol will be added later.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,7 +1,168 @@
...
+    ///
+    /// # Arguments
+    /// * `queue_index` - Index of the queue to modify.
+    /// * `fd` - EventFd that will be signaled from guest.
+    fn set_vring_kick(&self, queue_index: usize, fd: &EventFd) -> Result<()>;
+}
+
+/// An interface for controlling vsock vHost backend devices.
+#[cfg(feature = "vhost-vsock")]
+pub trait VhostVsock<'a>: VhostDevice<'a> {
```
Comment (sboeuf) [inline src/lib.rs]: We could define this one in its own `vsock.rs` file.

And also, why do you need this to be a `trait`? Do you expect multiple implementation of this?

File: sys_util/src/eventfd.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+use std::os::unix::io::{AsRawFd, FromRawFd, RawFd};
+use std::{io, mem, result};
+
+use libc::{c_void, dup, eventfd, read, write, EFD_NONBLOCK};
+
+/// A safe wrapper around a Linux eventfd (man 2 eventfd).
+///
+/// An eventfd is useful because it is sendable across processes and can be used for signaling in
+/// and out of the KVM API. They can also be polled like any other file descriptor.
+pub struct EventFd {
```
Comment (sboeuf) [inline sys_util/src/eventfd.rs]: What's the status on this crate `vmm-sys-utils`? I'm asking because we cannot merge the PR as it is now, right? I don't think we want to pull `sys_utils/...` into the `vhost` crate.

File: src/vhost_kern/vhost_binding.rs [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+//
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+/* Auto-generated by bindgen then manually modified for simplicity */
+
+#![allow(non_upper_case_globals)]
```
Comment (sboeuf) [inline src/vhost_kern/vhost_binding.rs]: This is being discussed on the ML, and this is not clear yet, but it seems you will have to put those bindings in a separate crate.

File: src/lib.rs [hunk 4]
```diff
@@ -1,7 +1,168 @@
...
+/// Result of vHost operations
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// An interface for setting up vhost-based virtio devices.
+///
+/// Vhost-based virtio devices are different from regular virtio devices because the the vhost
+/// backend takes care of handling all the data transfer. The device itself only needs to deal with
+/// setting up the the backend driver and managing the control channel.
+#[cfg(feature = "vhost-device")]
+pub trait VhostDevice<'a>: std::marker::Sized {
```
Comment (sboeuf) [inline src/lib.rs]: Thinking about the future with `vhost-user`, do you expect `vhost-user` devices to implement this trait directly (given that it needs to be extended)? Or do you expect the creation of a new `trait` `VhostUserDevice` that would inherit from this trait?

File: sys_util/src/eventfd.rs (hunk 2 above)
Comment (jiangliu) [reply sys_util/src/eventfd.rs]: Yes, that's a problem. This crate depends on the vmm-sys-utils and vm-memory.

File: src/vhost_kern/vhost_binding.rs (hunk 3 above)
Comment (jiangliu) [reply src/vhost_kern/vhost_binding.rs]: I have replied to the ML thread. For the vhost case, I do feel it's better to include the binding into the crate with manual modifications.

File: src/lib.rs (hunk 4 above)
Comment (jiangliu) [reply src/lib.rs]: I think it would be better that vhost-user implements the VhostDevice trait, though we need to extend the VhostDevice trait. Otherwise VhostDevice trait is almost useless:)

File: src/lib.rs (hunk 1 above)
Comment (jiangliu) [reply src/lib.rs]: Sure, I will move it to vsock.rs.
The existing vsock implementation is used a sample, because net and blk is not ready yet.

File: src/vhost_kern/vhost_binding.rs (hunk 3 above)
Comment (sboeuf) [reply src/vhost_kern/vhost_binding.rs]: @sameo @bonzini @andreeaflorescu @zachreizner need feedback here :)

File: src/lib.rs (hunk 4 above)
Comment (bonzini) [reply src/lib.rs]: Yes, I would expect at least two implementations for vhost-kernel and vhost-user respectively.

Comment (sboeuf) [reply src/lib.rs]: `VhostDevice` trait is needed if you want to implement both `vhost_vsock` and `vhost_net`, right?

But in case of vhost-user, I'm think we should create a trait extending `VhostDevice`, as it will be cleaner and clearer from any vhost-user device to implement a trait specific to vhost-user.

Comment (bonzini) [reply src/lib.rs]: Maybe rename this to VhostBackend?

**Comment by sboeuf** [comment]
@jiangliu @bonzini @andreeaflorescu what's the status here?
I'm very interested in seeing this crate moving forward in order to base some vhost-user implementations off of it.

**Comment by jiangliu** [comment]
> @jiangliu @bonzini @andreeaflorescu what's the status here?
> I'm very interested in seeing this crate moving forward in order to base some vhost-user implementations off of it.

I'm refine the crate to better support reconnection and almost ready, plan to send out early next week. I have also decoupled Unix domain socket client/server from host-user master/slave roles to enable the       use case that the vhost-user master acts as servers.

File: Cargo.toml [hunk 5]
```diff
@@ -1,8 +1,23 @@
 [package]
-name = "vhost"
+name = "vhost_rs"
```
Comment (sboeuf) [inline Cargo.toml]: Why did you change for `vhost_rs`?

Comment (jiangliu) [reply Cargo.toml]: it conflicts with firecracker/devices/src/virtio/vhost, so a temp solution to workaround the conflict.

File: src/vhost_kern/vhost_binding.rs (hunk 3 above)
Comment (andreeaflorescu) [reply src/vhost_kern/vhost_binding.rs]: From the coverage test perspective is better to have the bindings in a different crate. In Firecracker we are excluding auto-generated files when computing the coverage. If we have them in the same crate it means we will have to make changes to the test_coverage.py so that it is more  flexible and we can exclude independent files depending on the repository being tested. The context is the plan to have common integration tests in one repository and all other repositories to consumes them.

But if we don't use most of the things that are auto-generated then maybe we should just have a minimal vhost_bindings file inside the repsository. I don't have any preference between these two.

**Comment by sboeuf** [comment]
@jiangliu the PR LGTM! Are we waiting for review from some folks or do you still have some work planned on this PR?
/cc @andreeaflorescu @bonzini @sameo

**Comment by jiangliu** [comment]
> @jiangliu the PR LGTM! Are we waiting for review from some folks or do you still have some work planned on this PR?
> /cc @andreeaflorescu @bonzini @sameo
Currently waiting reviews, and then need to refine documentation and other required of the rust-vmm project.

File: src/vhost_kern/vhost_binding.rs (hunk 3 above)
Comment (jiangliu) [reply src/vhost_kern/vhost_binding.rs]: Good point about the test_coverage.py. Could we solve this issue by adding more test cases?

File: Cargo.toml (hunk 5 above)
Comment (bonzini) [reply Cargo.toml]: What is the issue exactly?

File: src/backend.rs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    pub mmap_offset: u64,
+    /// Optional file diescriptor for mmap
+    pub mmap_handle: RawFd,
+}
+
+/// An interface for setting up vhost-based backend drivers.
+///
+/// Vhost-based virtio devices are different from regular virtio devices because the the vhost
+/// backend takes care of handling all the data transfer. The device itself only needs to deal with
+/// setting up the the backend driver and managing the control channel.
```
Comment (stefanha) [inline src/backend.rs]: This comment is not generally applicable and relates to vhost-net only.  It may cause confusion.

The comment is essentially saying that in vhost-net the rx/tx virtqueues are handled by vhost and the ctrl virtqueue is handled by the VMM.

In the general case the responsibility of which virtqueue is handled by vhost and which is handled by the VMM is decided when the first vhost implementation of a device is designed.  Usually the dataplane is handled by vhost and anything else is handled by the VMM, but this is not a strict rule or enforced in any way.

I think a clearer way of explaining vhost is that a vhost device is a subset of a virtio devices.  Sometimes all virtqueues but often just those considered the dataplane.  The purpose of vhost is to implement a subset of a virtio device's functionality outside the VMM process.

File: src/lib.rs [hunk 7]
```diff
@@ -1,7 +1,109 @@
...
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD file.
+
+//! Virtio Vhost Backend Drivers
+//!
+//! Virtio devices use virtqueues to transport data efficiently. Virtqueue is a set of three
+//! different single-producer, single-consumer ring structures designed to store generic
+//! scatter-gather I/O.
```
Comment (stefanha) [inline src/lib.rs]: This is outdated as of VIRTIO 1.1.  The new Packed Virtqueue layout is a single ring.

This comment could be updated to:
Virtqueues are often implemented as VIRTIO 1.1 Split Virtqueues with a set of three ...

Eventually this crate should support Packed Virtqueues too for better performance.

**Comment by stefanha** [comment]
> @sboeuf @jiangliu It seems that the item max_queue_num in node should not be configured by VMM. max_queue_num is obtained from slave by calling get_queue_num, to indicate the maximum queue number slave could support. It could be a gatekeeper to check if VMM setup the queues more than slave cloud handle. I see the set_vring_* functions is handling like this. Thus, I thought the support added here should be removed. Correct me if I'm wrong :-P

This appears to be wrong.  Here is what I've found:

In vhost-user "multiple virtqueues" and "multiqueue" are two separate concepts.  Devices with multiple virtqueues have always been possible.  From the beginning of vhost the vhost-net device had an rx virtqueue and a tx virtqueue.  There was just no way for the master to ask the slave if more virtqueues are supported.

Devices with fixed virtqueue layouts are NOT required to advertise VHOST_USER_PROTOCOL_F_MQ because the VMM already knows the number of virtqueues.

When multiqueue devices were implemented it was necessary to extend the protocol so that the slave could report how many virtqueues are available.  Multiqueue devices have a variable number of virtqueues that depends on the capabilities of the device (i.e. the vhost-user slave).

I think that ideally all vhost-user device backends should implement VHOST_USER_PROTOCOL_F_MQ and VHOST_USER_GET_QUEUE_NUM so the master can fetch max_queue_num from the vhost device backend.

However, according to the vhost-user specification and how existing devices are implemented, it is not possible to rely on this today.  Here is how it works:

1. The VMM configures max_queue_num based on the number of queues offered by its virtio device (NOT the vhost device).
2. If the vhost-user device backend supports VHOST_USER_GET_QUEUE_NUM then max_queue_num can be fetched and updates its value.
3. If the VMM user explicitly requested a specific multiqueue number (e.g. on the command-line or through a management API) then device creation fails if this number exceeds max_queue_num (e.g. "Cannot create virtio-net device with 1024 virtqueues, vhost-user slave only supports 64").

Therefore Sebastien's patch or something similar is necessary.  The current master code conflates multiqueue with multiple virtqueues when they are actually separate concepts.

I have sent QEMU patches to make libvhost-user always advertise VHOST_USER_PROTOCOL_F_MQ but please do not rely on this.

**Comment by bjzhjing** [comment]
@stefanha There is one thing I would to confirm with you, that max_queue_num represents the queue pair number, rather than the total number of virtqueues, right?

**Comment by stefanha** [comment]
> @stefanha There is one thing I would to confirm with you, that max_queue_num represents the queue pair number, rather than the total number of virtqueues, right?

max_queue_num is the number of virtqueues.

**Comment by jiangliu** [comment]
> Great, I'll let you know when this will be solved on the virtio-fs side of things.

I plan to integrate this workaround into vhost with comments stating that it's a temporary workaround.

File: src/lib.rs (hunk 7 above)
Comment (jiangliu) [reply src/lib.rs]: The comment has been updated as:
Virtio devices use virtqueues to transport data efficiently. The first generation of virtqueue
is a set of three different single-producer, single-consumer ring structures designed to store
generic scatter-gather I/O. The virtio specification 1.1 introduces an alternative compact
virtqueue layout named "Packed Virtqueue", which is more friendly to memory cache system and
hardware implemented virtio devices. The packed virtqueue uses read-write memory, that means
the memory will be both read and written by both host and guest. The new Packed Virtqueue is 
preferred for performance.

**Comment by jiangliu** [comment]
>In vhost-user "multiple virtqueues" and "multiqueue" are two separate concepts. Devices with multiple >virtqueues have always been possible. From the beginning of vhost the vhost-net device had an rx >virtqueue and a tx virtqueue. There was just no way for the master to ask the slave if more virtqueues are >supported.
Thanks for the explanation, it's easy to mix up these two concepts.
I will carry the patch from @sboeuf in next PR.

File: src/backend.rs (hunk 6 above)
Comment (jiangliu) [reply src/backend.rs]: The comment has been updated as:
/// Vhost devices are subset of virtio devices, which improve virtio device's performance by
/// delegating data plane operations to dedicated IO service processes. Vhost devices use the
/// same virtqueue layout as virtio devices to allow vhost devices to be mapped directly to
/// virtio devices.
/// The purpose of vhost is to implement a subset of a virtio device's functionality outside the
/// VMM process. Typically fast paths for IO operations are delegated to the dedicated IO service
/// processes, and slow path for device configuration are still handled by the VMM process. It may
/// also be used to control access permissions of virtio backend devices.

**Comment by sameo** [comment]
This is superseded by PR #9
